pkg/ddl/partition.go

```diff
@@ -270,20 +265,36 @@ func alterTableLabelRule(schemaName string, meta *model.TableInfo, ids []int64)
 	return false, nil
 }
 
-func alterTablePartitionBundles(t *meta.Mutator, tblInfo *model.TableInfo, addingDefinitions []model.PartitionDefinition) ([]*placement.Bundle, error) {
+func alterTablePartitionBundles(t *meta.Mutator, tblInfo *model.TableInfo, addingDefinitions []model.PartitionDefinition) (bool, error) {
 	var bundles []*placement.Bundle
 
-	// tblInfo do not include added partitions, so we should add them first
+	// We want to achieve:
+	// - before we do any reorganization/write to new partitions/global indexes that the placement rules are in-place
+	// - not removing any placement rules for removed partitions
+	// - not leaving anything in case of failure/rollback!
+	// So we will:
+	// 1) First write the new bundles including both new and old partitions,
```

Comment:
Why the new bundles should include old partitions? The old bundles have already defined these rules and still take effects here.

Reply:
Oh, so if the Bundle is overwritten, the older bundle rules will still be in effect?
Like if the first bundle had rules for table id 111, part id 112 and 113, and the new bundle would have table id 111, partition id 113 and 114, would part id 112 still be placed as the old bundle rules or would it be placed without any rules?

Reply:
> or would it be placed without any rules?

This, I believe.

```
// it will filter the rules depends on the interval index override in the same group or the
// group-index override between different groups
// For example, given rules:
// ruleA: group_id: 4, id: 2, override: true
// ruleB: group_id: 4, id: 1, override: true
// ruleC: group_id: 3
// ruleD: group_id: 2
// RuleGroupA: id:4, override: false
// RuleGroupB: id:3, override: true
// RuleGroupC: id:2, override: false
// Finally only ruleA and ruleC will be selected.
```

Reply:
A better example:
```
CREATE TABLE t (a int) partition by range (a) (partition p0 values less than (1000000), partition p1M values less than (2000000)) placement policy pp1;
ALTER TABLE t REORGANIZE PARTITION p1M INTO (PARTITION p1M values less than (2000000), partition p2M values less than (3000000));
```
During the DDL we would still want the original p1M to follow the table level rule, so we keep it in the bundle.
When the final bundles will be created it will not be included.
Since the table keeps its table id (for REORGANIZE PARTITION) there are no other bundles that covers the inherited table's placement rules for the original partition p1M, which means we need to cover it during the time of the DDL, since it can still be accessed (and double written).

Another alternative would be to create partition level bundles for the old replaced partitions, and let them be removed by GC later, but I think that would qualify for a separate PR.

Reply:
I think the rule bundle check improvements would be better addressed in a separate PR, I will update the check and add a unit test here, but when looking at the code in PD, it looks like the example in the comment for [`prepareRulesForApply()`](https://github.com/tikv/pd/blob/52b42b25ba8f8f75943200b76d48565b8aace2b8/pkg/schedule/placement/rule.go#L182-L190) is wrong, if I'm reading the code correctly, I would expect it to return 'ruleC and ruleD`, since ruleB would override ruleA, and ruleC would override all previous rules, due to its group has override set, and finally ruleD would just be added.

And it does not check any key ranges, neither in this function or in `checkApplyRules()`, which only checks that there are at least one leader or voter, and max one leader.

Reply:
I have changed the logic to only check bundles one by one, and tried to mimic the logic of PDs `prepareRulesForApply()` and `checkApplyRules()`, for any other check I created https://github.com/pingcap/tidb/issues/57693 as a follow up.